code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/configuration/ConfigurationExtension.kt

```diff
@@ -0,0 +1,575 @@
+    private val launchRulesEvaluator: LaunchRulesEvaluator
+    private val configurationStateManager: ConfigurationStateManager
+    private val configurationRulesManager: ConfigurationRulesManager
+    private val retryWorker: ScheduledExecutorService
+    private var retryConfigurationCounter: Int = 0
+    private var retryConfigTaskHandle: Future<*>? = null
+
+    constructor(extensionApi: ExtensionApi) : this(extensionApi, ServiceProvider.getInstance())
+
+    /**
+     * Exists only for cascading components for dependency injection.
+     */
+    private constructor(
+        extensionApi: ExtensionApi,
+        serviceProvider: ServiceProvider
+    ) : this(
+        extensionApi, serviceProvider,
+        AppIdManager(serviceProvider.dataStoreService, serviceProvider.deviceInfoService),
+        CacheManager(serviceProvider.deviceInfoService),
+        LaunchRulesEvaluator("Configuration", LaunchRulesEngine(extensionApi), extensionApi),
+        Executors.newSingleThreadScheduledExecutor()
+    )
+
+    /**
+     * Exists only for cascading components for dependency injection.
+     */
+    private constructor(
+        extensionApi: ExtensionApi,
+        serviceProvider: ServiceProvider,
+        appIdManager: AppIdManager,
+        cacheFileService: CacheFileService,
+        launchRulesEvaluator: LaunchRulesEvaluator,
+        retryWorker: ScheduledExecutorService
+    ) : this(
+        extensionApi,
+        serviceProvider,
+        appIdManager,
+        cacheFileService,
+        launchRulesEvaluator,
+        retryWorker,
+        ConfigurationStateManager(
+            appIdManager,
+            cacheFileService,
+            serviceProvider.networkService,
+            serviceProvider.deviceInfoService,
+            serviceProvider.dataStoreService
+        ),
+        ConfigurationRulesManager(
+            launchRulesEvaluator,
+            serviceProvider.dataStoreService,
+            serviceProvider.deviceInfoService,
+            serviceProvider.networkService,
+            cacheFileService
+        )
+    )
+
+    @VisibleForTesting
+    internal constructor(
+        extensionApi: ExtensionApi,
+        serviceProvider: ServiceProvider,
+        appIdManager: AppIdManager,
+        cacheFileService: CacheFileService,
+        launchRulesEvaluator: LaunchRulesEvaluator,
+        retryWorker: ScheduledExecutorService,
+        configurationStateManager: ConfigurationStateManager,
+        configurationRulesManager: ConfigurationRulesManager
+    ) : super(extensionApi) {
+        this.serviceProvider = serviceProvider
+        this.appIdManager = appIdManager
+        this.cacheFileService = cacheFileService
+        this.launchRulesEvaluator = launchRulesEvaluator
+        this.retryWorker = retryWorker
+        this.configurationStateManager = configurationStateManager
+        this.configurationRulesManager = configurationRulesManager
+
+        loadInitialConfiguration()
```

Comment:
Shared states should not be set from constructor. This is a limitation with our Extension APIs. We should move the first shared state update to `onRegistered()` method.

Reply:
nvm. I noticed the logic in `onRegistered()` method